### PR TITLE
[Fix] Bug 34: Monitor files at Windows root directory

### PR DESCRIPTION
**Pull request by TinaMor**
## PR Description
This PR allows us to monitor files in the root directory e.g. C:\\

### Solution
For root directory, do not prepend, otherwise, prepend "\\?\" to the directory path. We validate the config file to check that when the root directory is provided, `includeSubdirectories` must be `false`. Otherwise, the appliation terminates with an error.

image

**TL;DR**

By default, the name is limited to MAX_PATH characters. To extend this limit to 32,767 wide characters, we prepend "\\?\" to the path in [LogFileMonitor.cpp](https://github.com/microsoft/windows-container-tools/blob/a47d40e76f63d8ad0f890b261783b862616fcd6a/LogMonitor/src/LogMonitor/LogFileMonitor.cpp#L64). Prepending the string "\\?\" does not allow access to the root directory. [Maximum Path Length Limitation](https://learn.microsoft.com/en-us/windows/win32/fileio/maximum-file-path-limitation?tabs=registry)
In editions of Windows before Windows 10 version 1607, the maximum length for a path is MAX_PATH, which is defined as 260 characters. In later versions of Windows, changing a registry key or using the Group Policy tool is required to remove the limit. See Maximum Path Length Limitation for full details.
There is no limit on the depth of a file or directory. With the standard API you can create as many folders as you want until the actual length of the expanded path exceeds 32,767 characters. The math on that is 2¹⁵ minus 1. The limitation of 32,767 doesn’t come from NTFS. It comes from the Win32 API.

### References
1.	[FindFirstFileW function (fileapi.h)](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-findfirstfilew)
2.	[Naming Files, Paths, and Namespaces](https://learn.microsoft.com/en-us/windows/win32/fileio/naming-a-file)
3.	[Maximum Path Length Limitation](https://learn.microsoft.com/en-us/windows/win32/fileio/maximum-file-path-limitation?tabs=registry)

### CODE REVIEW

File: LogMonitor/src/LogMonitor/ConfigFileParser.cpp [hunk 1]
```diff
@@ -640,6 +652,38 @@ AddNewSource(
...
+    const std::wstring key = L"directory";
+
+    if (!Utility::ConfigAttributeExists(Attributes, JSON_TAG_DIRECTORY) ||
+        !Utility::ConfigAttributeExists(Attributes, JSON_TAG_INCLUDE_SUBDIRECTORIES))
+    {
+            return true;
+    }
+
+    std::wstring directory = *(std::wstring *)Attributes[JSON_TAG_DIRECTORY];
+    const bool includeSubdirectories = *(bool *)Attributes[JSON_TAG_INCLUDE_SUBDIRECTORIES];
```
Comment (iankingori) [inline LogMonitor/src/LogMonitor/ConfigFileParser.cpp:671]: Is includeSubDirectories a required field and what does it default to  if nothing is passed?

Comment (TinaMor) [reply LogMonitor/src/LogMonitor/ConfigFileParser.cpp:671]: `includeSubDirectories` is an optional field. It defaults to `false` when not provided. [LoggerSettings.h#l25](https://github.com/microsoft/windows-container-tools/blob/a47d40e76f63d8ad0f890b261783b862616fcd6a/LogMonitor/src/LogMonitor/Parser/LoggerSettings.h#L257C5-L257C40)

File: LogMonitor/LogMonitorTests/LogMonitorTests.cpp [hunk 2]
```diff
@@ -9,7 +9,7 @@
 #include "../src/LogMonitor/EtwMonitor.cpp"
 #include "../src/LogMonitor/EventMonitor.cpp"
 #include "../src/LogMonitor/JsonFileParser.cpp"
-#include "../src/LogMonitor/FileMonitor/Utilities.cpp"
+#include "../src/LogMonitor/FileMonitor/FileMonitorUtilities.cpp"
```
Comment (CharityKathure) [inline LogMonitor/LogMonitorTests/LogMonitorTests.cpp:12]: Why is this utility file specific to files? can the functions here be moved in LogFileMonitor.cpp (if they are specific file monitoring) or the Utility.cpp file (if they can be utilized throughout the tool)?

**Comment by bobsira** [comment]
The looks good to me. I've noticed a few issues while try to test your code.

1. If I begin monitoring the C: without having a log file and create the logfile later on I keep getting `LogFilesChangeHandler: Timer event` even after the log file is created and I'm pushing data to that log file.
2. If I begin monitoring the C: with a log file already there I still see the `LogFilesChangeHandler: Timer event`

Is this the expected behavior and I'm not sure if it is related to this PR as well?

logfilemon

File: LogMonitor/LogMonitorTests/LogMonitorTests.cpp (hunk 2 above)
Comment (TinaMor) [reply LogMonitor/LogMonitorTests/LogMonitorTests.cpp:12]: The methods are only specific to LogFileMonitor. Moving them to LogFileMonitor.cpp will make the file too long a difficult to manage. Hence the need to seperate the utilities methods

**Comment by TinaMor** [comment]
@bobsira Please share the config file and the container image you are using.

> The looks good to me. I've noticed a few issues while try to test your code.
> 
> 1. If I begin monitoring the C: without having a log file and create the logfile later on I keep getting `LogFilesChangeHandler: Timer event` even after the log file is created and I'm pushing data to that log file.
> 2. If I begin monitoring the C: with a log file already there I still see the `LogFilesChangeHandler: Timer event`
> 
> Is this the expected behavior and I'm not sure if it is related to this PR as well?
> 
> logfilemon

**Review by bobsira — APPROVED** [review]

**Pull request merged**